tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15360,8 +15360,30 @@ void DiagnoseNodeEntry(Sema &S, FunctionDecl *FD, HLSLShaderAttr *Attr) {
 
 void DiagnoseEntry(Sema &S, FunctionDecl *FD) {
   auto Attr = FD->getAttr<HLSLShaderAttr>();
-  if (!Attr)
-    return;
+  if (!Attr) {
+    // if this is the Entry FD, then add the target profile
+    // shader attribute to the FD and carry on with validation
+    // otherwise, with no shader attribute, just return
+    const std::string &EntryPointName = S.getLangOpts().HLSLEntryFunction;
+    if (EntryPointName.empty()) {
+      return;
+    }
+
+    
+    // if this FD isn't the entry point, then there's no
+    // shader attribute to work with, so just return
+    if (EntryPointName != FD->getIdentifier()->getName()) {
+      return;
+    }
+
+    std::string profile = S.getLangOpts().HLSLProfile;
+    const ShaderModel *SM = hlsl::ShaderModel::GetByName(profile.c_str());
+    const llvm::StringRef fullName = ShaderModel::FullNameFromKind(SM->GetKind());
```

Comment:
The attribute should only be added when we have a valid shader target (as opposed to a library target, for instance).  If you follow my other suggestion about returning an empty `StringRef()`, this can be accomplished by only adding the attribute when the `fullName` returned is not empty.